### DISCUSSION TITLE
Plex Music Provider

### DISCUSSION THREAD

**Discussion by marcelveldt** [music-assistant/support#1168] [Music providers]
Music Assistant now has support for using Plex (MusicLibrary) as a Music Assistant provider

Documentation https://music-assistant.io/music-providers/plex/

**Comment by jeroenterheerdt** [comment]
I am getting a timeout when setting this up. I verified the X-Plex-Token is correct and I set up the Server as the name of the server in the UI in Plex. Or should it be something else? I am wondering if this is dependent on network setup? @micha91 ?

**Reply by micha91** [reply]
The plexapi library tries all known addresses of a server and selects the best one. It is possible that the MA timeout kills the plex provider before it was able to try all known addresses. This will change with the next beta, as the timeout on the plexapi side will be reduced to 10 seconds (now it is most likely 60 seconds) and MA has a timeout of 30 seconds

**Reply by jeroenterheerdt** [reply]
thanks @micha91! So what "know addresses" is this trying? It doesn't seem accept public or local IP addresses, so what else is there?

**Reply by marcelveldt** [reply]
In the Plex server settings there's a setting where you can configure that

**Reply by jeroenterheerdt** [reply]
so this is what my settings look like - I tried all the items I covered in the screenshot: the name of the server, the internal IP, the public IP. None of them works. What should I be using?

![image](https://user-images.githubusercontent.com/8188990/228931117-52c83de5-720c-45b5-86f4-03eb02063449.png)

**Reply by marcelveldt** [reply]
Wait for the next beta and try again, it has a possible fix for this issue

**Reply by jeroenterheerdt** [reply]
awesome, will do. Thanks both!

**Comment by jeroenterheerdt** [comment]
just wanted to let you know that it looks like this music provider works now - I was able to configure it without any issue. Awesome job!

**Comment by ashbrnet** [comment]
this is working however a cosmetic suggestion on the page to add players, the link myplex.tv invokes a pop up which most people have blocked and it will 999 timeout error, add some text to say it will open in pop up window or similar

**Reply by OzGav** [reply]
Thoughts @micha91 ?

**Reply by micha91** [reply]
I like the idea, but we should add this for all providers, which use external authentication - there will always be the same issue, I believe. I will talk to the other devs and we will discuss it

**Comment by Benny12323** [comment]
Hi guys,
thank you so much for this! When I try to link my Plex Account I get an error:
`'PlexClient' object has no attribute 'library'
`
It occurs after being redirected from the login at myplex.tv Any ideas?

**Reply by RJ-Make** [reply]
> Hi guys, thank you so much for this! When I try to link my Plex Account I get an error: `'PlexClient' object has no attribute 'library' ` It occurs after being redirected from the login at myplex.tv Any ideas?

Do you get the full log-in page or just one with an email address?  If it's that latter, then use the "Use a different email address" option.

**Reply by Benny12323** [reply]
@RJ-Make I tried that, but the behaviour doesn't change.  The logs don't give any more hints either...

**Reply by micha91** [reply]
Hi, there currently is a bug in the authentication. Currently we do not filter the list of connected resources and due to that we also try to connect to players and not only to servers. I already implemented a fix, which should be available soon

**Comment by petermcc1** [comment]
hi,
I am having an issue with Plex and MA not showing playlists that I create (in Plex). I've tried the re-sync, but the new playlists still do not appear in MA. Is there something I need to do in Plex so new playlists appear in MA?

Other than this issue, this is an incredible platform. It does what so many companies have been charging tons of $$$ but have never quite been able to pull off. So a HUGE congrats to everyone that made MA happen. Put me down for a t-shirt when they are available!!

thx.

**Reply by micha91** [reply]
Did you create a Playlist or a collection? Right now we only support Playlists in MA and not Collection. If the issue is on Playlists it's a bug and we should investigate further. In that case some logs would be really helpful

**Comment by bj61251** [comment]
I'm having a similar issue with Playlists not syncing from Plex. I've been able to manually choose "Add to Library" via the Browse section for new playlists that don't appear right away, but if I simply modify a playlist in Plex (or for 'smart' playlists that auto update based on filter criteria), I can't seem to do anything to get it to update in Music Assistant (even the Browse > Plex Media Server Library > Playlists version). I've tried removing/re-adding, syncing Plex, restarting the server container, etc... The only thing that appears to work is to create a separate copy of the playlist with a new name, which mostly defeats the purpose. Is there any way to actually force the playlist to re-sync with Plex server?

**Reply by OzGav** [reply]
Plex is getting some attention at the moment. Can you open an issue about this so we can follow it up.

**Comment by DarkAllMan** [comment]
When I want to add a Plex Library I get this error: library_id has unexpected type: <class 'list'>

Solved it by deselecting and reselecting the library in the dropdown.

**Comment by antialias** [comment]
Where is "the server field"? This is all I see in the music assistant provider setup page.
<img width="1440" alt="Screenshot 2024-01-18 at 11 34 37" src="https://github.com/music-assistant/hass-music-assistant/assets/447517/54253974-4219-4aea-b610-f7abf8cd6b63">

**Reply by OzGav** [reply]
Please wait for the next beta. There is an improvement coming for the initial setup.

**Comment by shedman214** [comment]
Getting these errors, is there something i'm doing wrong? 
![image](https://github.com/music-assistant/hass-music-assistant/assets/32603273/d78e12af-7fe0-4d57-8890-51568335b03d)
![image](https://github.com/music-assistant/hass-music-assistant/assets/32603273/bc45ee8e-ab12-4497-9275-36561c2fee1e)

**Reply by OzGav** [reply]
@jozefKruszynski ?

**Reply by jozefKruszynski** [reply]
It looks like you're trying to connect using the error messages as the IP and port.
You need to replace those error messages with your server's actual IP and port.

**Reply by shedman214** [reply]
@jozefKruszynski  I have put in the local ip and port. I get that message when i enter it. I also have enabled the GDM on plex and still does not work

**Reply by OzGav** [reply]
Can you try clearing your browser cache?

**Reply by shedman214** [reply]
I have cleared the cache for 7 days and still same error

**Reply by OzGav** [reply]
Have you tried a different browser or device?

**Reply by shedman214** [reply]
Yes i have tried nabu casa, my local HA, chrome and edge. no luck
![image](https://github.com/music-assistant/hass-music-assistant/assets/32603273/408abc60-9af0-4e08-906b-0ed6f93e3762)

**Reply by jozefKruszynski** [reply]
Can you please add your steps exactly,
I have tried to reproduce the issue and I'm unable to do so.

Is your Plex server your own and on your local network?
How is it running? i.e. docker, lxc, windows etc..?
Have you tried restarting the Plex server? 

Can you see if there are any logs in the system log related to this error and post them please.

**Reply by shedman214** [reply]
First step is I go to authenticate the plex server and I get an error. I then try the local ip way and get "Discovery failed, please add IP manually"
I have my plex media server on my unraid server that uses docker. I'm not seeing anything in the logs unless i'm not looking in the right spot?

**Reply by shedman214** [reply]
My apologies, I found the logs 
[d5369777_music_assistant_beta_2024-01-25T14-27-16.383Z.log](https://github.com/music-assistant/hass-music-assistant/files/14053334/d5369777_music_assistant_beta_2024-01-25T14-27-16.383Z.log)

**Reply by jozefKruszynski** [reply]
Hi, I updated the instructions for the provider at the top of this discussion, can you try to follow the new instructions and see if it resolves your issue please

**Reply by jozefKruszynski** [reply]
@shedman214 Any update?

**Reply by shedman214** [reply]
I have followed the instructions up top and still get errors for both ways. Should I uninstall everything and start over from scratch?

**Reply by OzGav** [reply]
Have you tried just removing the Plex provider? If not, then yes do that

**Reply by jozefKruszynski** [reply]
> I have followed the instructions up top and still get errors for both ways. Should I uninstall everything and start over from scratch?

What do you mean both ways, there is only really one way. The only difference is whether or not the discovery works.
All the discovery does though is fill in those text fields with the appropriate values prior to you clicking the auth button.

**Reply by shedman214** [reply]
I get this error when i try to authenticate plex. 
![image](https://github.com/music-assistant/hass-music-assistant/assets/32603273/5c1dc5c3-d531-4529-9036-de931d25c363)

**Reply by shedman214** [reply]
[d5369777_music_assistant_beta_2024-02-05T13-21-03.073Z.log](https://github.com/music-assistant/hass-music-assistant/files/14166203/d5369777_music_assistant_beta_2024-02-05T13-21-03.073Z.log)

**Reply by jozefKruszynski** [reply]
I'm not sure what to say here, the image suggests that you are simply not adding the and port to the relevant fields

**Reply by shedman214** [reply]
where do i add them because i am adding them here but its not accepting when i add them manually? 
![image](https://github.com/music-assistant/hass-music-assistant/assets/32603273/4e57c583-ce41-43e2-850d-b7d55beec0be)

**Reply by shedman214** [reply]
https://logpaste.com/Nhf6u7tY

**Reply by jozefKruszynski** [reply]
Can you please uninstall both the music assistant addon and the integration and try again.
It looks like you are doing everything correctly so I think the reinstall is the best option.
Also please do a reboot of HA between uninstall and reinstall of MA

**Reply by shedman214** [reply]
I deleted and restarted—still the same error. When i first ever installed it it worked. but after some updates, it does not work anymore. is it a firewall issue?

**Comment by jnathalia** [comment]
All is working great...

But I have 2 friends, with external IP address, trying to let them add my Plex music server but this fails. Is that intentionally or is it something else? When "Authenticate on myplex.tv" I log in with my friends account and it is successful at the new pop-up Plex webpage, but it keeps on trying to finish the add procedure at HA/MA for the "Add music provider".

**Reply by RJ-Make** [reply]
I'm probably missing something here, but wouldn't you add them to **Your Plex Server** and then share your music collection with them? When they add **Their Plex** information into their MA instance they would have access to the shared music on MA.

**Reply by jnathalia** [reply]
I see what you are saying, but those 2 friends don't have their own server. So, when filling in local server IP this doesn't make sense.

We have done the following:
- My 2 friends created their own account at app.plex.tv
- Then I have searched for them and become friends
- At my server I have allowed them to get access to my Plex server and TV tuner
- Now they are able to play anything they like directly from my server

This is all fine and via the Plex app on iPhone, browser or TV-app flawless.

Now they go to MA and fill in the local host. There isn't so they fill my DNS name (or external IP) and they sign in with their own Plex account. This is successful, however on the original page of HA it keeps on trying... 
Even with my account, I cannot complete the procedure successful. 

But bottomline, they don't have their own Plex server, they do have their own Plex account...

**Reply by RJ-Make** [reply]
Ah I see, that is indeed what I was missing. 

The only way I see this working without you using something like Twingate (Cloudflared also but you may run into bandwidth bans) to allow tunneled access to you MA port, is to have them install the server software on an always-on-machine/device.

**Reply by jozefKruszynski** [reply]
I guess when I changed the auth process I wasn't thinking about anything other than local servers. I'll try to look into re-adding remote server options, but there are many other things to look at before that in other areas of the app.
I can't promise anything I'm afraid

**Reply by jnathalia** [reply]
Great, as long as this problem is acknowledged, I will wait until it's there.

**Reply by OzGav** [reply]
You don’t mention if this music you are sharing is your own creation so just to make it clear we will not facilitate anything which violates the Plex terms of service or copyright law. https://www.plex.tv/en-au/about/privacy-legal/plex-terms-of-service/
> Third-party Content. Third-party content displayed or accessible through the Plex Solution is protected by copyright and other intellectual property law and belongs to the respective owner. Use of the third-party content is subject to the terms of use of the third party providing such content. This TOS does not grant you the right to copy, distribute, prepare derivative works, publicly display, or otherwise use any third-party content. You are expressly prohibited from engaging in or facilitating the unauthorized sharing or distribution of third-party content.

**Comment by Paul-Vdp** [comment]
Not sure if this problem is restricted to the Plex music provider only - but that's the only one I use.
I found out that the player attribute 'media_album_name' acts very erratic in case of a (plex) playlist being started.
Whereas when playing a classic album its name is duly repeated in the 'media_album_name', when starting a playlist I get something like '/library/metadata/20478' in there. To which I must add that the correct name (of the playlist) might appear after playing for some time - purely at random it seems.

Rather annoying, because one of my main music automations is based on this album name attribute :-(

**Reply by OzGav** [reply]
In case you aren’t aware the Plex provider is not currently maintained. We recommend you share your music directly with MA. Sorry.

**Comment by Paul-Vdp** [comment]
Well, I was not aware of that. Bad news that is.
I guess with 'share your music directly with MA' you mean using the Filesystem provider ?
How confident are you that the same problem/error does not exist with this provider and its playlists ?

**Reply by OzGav** [reply]
Yes the filesystem provider. What problem are you referring to?

**Comment by Paul-Vdp** [comment]
The problem as desccribed in my original message (with playlists and the player attribute 'media_album_name') ?

**Reply by OzGav** [reply]
No there are no issues like that with the native playlists. You will have to recreate it but then it is stored on your local disk and is named whatever you decide when it is created.

**Comment by kmanan** [comment]
I am unable to connect PLex as a music provider. These are the errors:

![image](https://github.com/music-assistant/support/assets/17659155/29ecc449-d39c-41e6-a9b4-f2d3e9057df7)
![image](https://github.com/music-assistant/support/assets/17659155/a555ecec-aa4e-42c4-8fe5-1745c839e732)
`2024-05-10 09:40:13.466 ERROR (MainThread) [music_assistant.webserver] Error handling message: config/providers/get_entries: Unable to retrieve Servers and/or Music Libraries
2024-05-10 09:40:33.881 WARNING (ThreadPoolExecutor-0_8) [py.warnings] /usr/local/lib/python3.12/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host 'plex.kamar-taj.net'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(

2024-05-10 09:40:33.991 WARNING (ThreadPoolExecutor-0_8) [py.warnings] /usr/local/lib/python3.12/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host 'plex.kamar-taj.net'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(

2024-05-10 09:40:34.015 WARNING (ThreadPoolExecutor-0_8) [py.warnings] /usr/local/lib/python3.12/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host 'plex.kamar-taj.net'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(

2024-05-10 09:40:34.052 ERROR (MainThread) [music_assistant.webserver] Error handling message: config/providers/get_entries: Unable to retrieve Servers and/or Music Libraries
2024-05-10 09:41:58.229 ERROR (MainThread) [music_assistant.webserver] Error handling message: config/providers/get_entries: Unable to retrieve Servers and/or Music Libraries
2024-05-10 09:46:59.247 ERROR (MainThread) [music_assistant.webserver] Error handling message: config/providers/get_entries: Unable to retrieve Servers and/or Music Libraries
2024-05-10 09:47:21.937 INFO (MainThread) [music_assistant.music] Sync task for Music Assistant completed
2024-05-10 09:48:58.598 ERROR (MainThread) [music_assistant.webserver] Error handling message: config/providers/get_entries: Unable to retrieve Servers and/or Music Libraries`

**Reply by bedar89** [reply]
I'm having these exact issues. Both with using myplex and locally. Does anybody know how to solve this?

**Comment by TyzzyT** [comment]
When playing music from Plex, it doesn't show up in the Plex server Now Playing dashboard. Is this intentional? When playing music in PlexAmp or via the Sonos app, it's showing the progress in the Plex Now Playing dashboard.
It would be nice if Plex would show what's playing right now, so I know when the server is being used and I wont restart. Extra plus: Plex can track the most popular tracks.

**Reply by OzGav** [reply]
Plex doesn't have a dedciated dev so feature requests are unlikely to get up. Having said that it might benefit from this if it gets enough votes https://github.com/orgs/music-assistant/discussions/632

**Reply by arsaboo** [reply]
@OzGav this is a slightly different issue. It is not about scrobbling to external services but keeping track of Plex play history (inside Plex).

**Comment by thegarywilder** [comment]
When trying to add Plex as a music provider, I can't save. The correct server IP and port are discovered; Authenticate Locally appears to work as the library is correctly populated; clicking on Save gives me a pop-up message with my local HA server IP and port as the title and "5" as the message.
If I attempt to select SSL before authenticating, I get a private certificate warning, or at least I did the first time.
If I then deselect Verify certificate under Advanced Settings, authentication appears to work, but SSL and Verify settings are reset and I'm back to getting "5" when I try to save.

This is the only HA log I can find:
Logger: py.warnings
Source: components/plex/server.py:317
First occurred: January 28, 2025 at 11:46:53 AM (2036 occurrences)
Last logged: 12:07:03 PM

/usr/local/lib/python3.13/site-packages/urllib3/connectionpool.py:1064: InsecureRequestWarning: Unverified HTTPS request is being made to host '192.168.xxx.xxx'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings warnings.warn(

Plex log:
DEBUG - Request: plaintext connection from 192.168.xxx.xxx:34652 rejected because secure connections are required

If I now change my Plex server setting from Required to Preferred for Secure Connections, I can save this Music provider. This should not be the case.

**Reply by Anto79-ops** [reply]
@thegarywilder <strike> have you figured this out? Im trying to bind my Plex Server from a Synolgy NAS but its not working. Locally does not work at all (error message), using MyPlex brings to the signon  page and says it sucessfully signed in but, does not re-redirect. I tried Required to Preferred (it was actually on preferred by default), no change.

Do you have any suggestions? thanks! </strike>

EDIT: ok, figured this out. Needed to add the HA music server to the list of allowed servers for Plex, and authenticated locally worked after that.

**Comment by tm05977** [comment]
Will play radio of a song or album be possible in the near future?
I can't get i to work, neither with plex nor with spotify

**Reply by OzGav** [reply]
What do you mean by “play radio of a song”? Do you mean use the MA radio mode functionality? If so the player provider has to support that. Spotify does but a recent change by Spotify means you can’t use your own client id and have that service.

**Reply by tm05977** [reply]
This is exactly what I meant, and Plex supports it.

I can see the "Start Radio" option, but when I click on it in a Plex album, I receive the following error:
`2025-02-26 09:01:36.899 ERROR (MainThread) [music_assistant.webserver] Error handling message: player_queues/play_media: list index out of range`

When attempting the same action on a Spotify album, I get a different error, but the result is the same, no sound:

`2025-02-26 09:04:02.119 ERROR (MainThread) [music_assistant.webserver] Error handling message: player_queues/play_media: recommendations not found`

**Reply by OzGav** [reply]
Firstly, make sure you are on 2.4.0. Next Spotify removed the ability to get recommendations if you use your own Client id. Plex is not really maintained at the moment so I didn't even know it supported the radio option. We would have to look into that but might be low priority depending on what else is going on

**Reply by tm05977** [reply]
I wasn’t on 2.4 before. Now, I no longer see the "Start Radio" option in either Plex or Spotify. I’ll wait patiently for this feature to be added.

**Reply by OzGav** [reply]
If you don't use your own client id (when you set up the Spotify provider)  then you will see it for spotify

**Comment by jlnbln** [comment]
Now that Music Assistant works with audiobooks I would like to add my Plex Audiobook library (I am using it with prologue on my phone). However when I add it, it is added as music and there is no way to change it to be recognized as audiobooks? Or is there? Thanks for your help.

**Reply by OzGav** [reply]
No sorry there is not. That functionality has to be added to each provider and there is no plex dev at the moment to do that work.

**Comment by juneix** [comment]
I hope Music Assistant can sync Plex's "Now Playing" status.
<img width="1984" alt="HapiGo_2025-04-17_09 20 45" src="https://github.com/user-attachments/assets/63ad3401-e124-4d36-baac-6318d1661895" />

**Reply by OzGav** [reply]
Please stop spamming all the threads with the same message

**Reply by juneix** [reply]
Oh, sorry, I'm not sure if some of the older related content is still being followed. I have already deleted some duplicate content

**Comment by OzGav** [comment]
Plex Connect support is available on 2.7.0 beta 16 and later

**Reply by arsaboo** [reply]
Is it different from the previous version of the Plex provider or part of #2512?

**Reply by OzGav** [reply]
It is a plugin different to the provider. https://beta.music-assistant.io/plugins/plex-connect/

**Reply by Anto79-ops** [reply]
Nice just tested it out, shared a speaker to Plex Amp via MA.

![Screenshot_20251125_211457_Plexamp~2.jpg](https://github.com/user-attachments/assets/5a16c87e-83ef-4bc0-ba0d-b3cbabcc817c)

**Comment by Slim1399** [comment]
Hi everyone,
First of all, great job !
You absolutely must try to integrate Plexamp; it will add another level to the experience or create an similar plugin to Plexamp capable of generating playlists as amazing as Plexamp's.

**Reply by OzGav** [reply]
This is literally what the post above this is about

**Comment by howels** [comment]
Please can you populate the "date added" field from Plex into the Provider data available to MA?
https://plexapi.dev/api-reference/content/get-items-in-a-collection#response-media-container-metadata-items-added-at

This would enable consistent album sorting by date added even if we have to reset the MA DB during testing, and also allow for sorting by date when collections are much older than the MA installation.

**Comment by SergeySSS3** [comment]
Hello. First of all, great work! I have a question about the display (sorting) of albums. When I go to Browse – Plex Media Server Library – Albums in the Home Assistant add-on, all album art appears correctly, just like in the Plex media server. However, if I go to Browse – Plex Media Server Library – Artist, incorrect data appears.
For example, in Plex, 2Cellos is represented by one artist, while in Music Assistant there are seven, with the album art only appearing on the first one.
<img width="1669" height="526" alt="120120" src="https://github.com/user-attachments/assets/03c76b33-3ade-4c9a-b357-7d42416c8505" />
<img width="1471" height="791" alt="130130" src="https://github.com/user-attachments/assets/979fcc28-dda1-4be0-ba31-a025a3444b3d" />

**Reply by OzGav** [reply]
Looks like inadequate or incorrect  tagging.

**Comment by SergeySSS3** [comment]
The artist and the artist of the album are the same everywhere, the comments, dates, and song titles differ.
<img width="733" height="499" alt="140140" src="https://github.com/user-attachments/assets/c1590bb3-6057-4785-9428-2812e09ef5aa" />

**Reply by OzGav** [reply]
I dont see any musicbrainz tags. But also I just looked again at your original image. That is browsing from Plex so the issue seems to be there. @anatosun

**Comment by SergeySSS3** [comment]
The problem is that with ordered tags (I mean, all files with one artist). The display in Plex and the display via music-assistant are different when sorted by artist.

**Reply by OzGav** [reply]
Are you referring to 2CELLOS now or something else?

**Reply by SergeySSS3** [reply]
Yes, that's right. But this is the case with many artists

**Reply by OzGav** [reply]
Ok so you say that the artist name is the “same everywhere” so where are those 2CELLOS feat. \<various\> coming from?

**Reply by howels** [reply]
Try using the filter for "album artist" - also make sure that when you look at each individual album in Plex it is showing the same data as in the tags. Plex can persist old metadata or fetch metadata from the internet which causes the new tags to be ignored.
Once Plex is 100% correct try deleting the artists from MA and syncing again.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
